### PR TITLE
Emulate/wrap upstream gsw.* submodules

### PR DESCRIPTION
**Pull request by DocOtak**
I've attempted "automagically" wrap all the upstream submodules in gsw.

Right now the one failing test was still failing on me with a wrong standard name attribute assertion:

```python

    def test_func_standard_module():
        """gsw can be used with modules, e.g. gsw.density"""
        ds = _create_ds()
        sigma0 = gsw.density.sigma0(SA=ds.SA, CT=ds.CT)
>       assert sigma0.attrs['standard_name'] == 'sea_water_sigma_t'
E       AssertionError: assert 'sea_water_absolute_salinity' == 'sea_water_sigma_t'
E         - sea_water_sigma_t
E         + sea_water_absolute_salinity

tests/test_gsw_xarray.py:29: AssertionError
```

But when I attempt to call the function manually, it has the correct attribute. So some more digging is needed before merging to understand this.

### CODE REVIEW

**Comment by DocOtak** [comment]
So the above error (and apparent heisenbug like nature of it), was caused by the different ways of accessing the submodules.

If you did a direct import such as:
```python
from gsw_xarray.density import sigma0
```
The expected wrapped function would be returned to you. However, if you did the following:
```python
import gsw_xarray
gsw_xarray.density.sigma0()
```
The ``gsw_xarray.density`` part is actually be an attribute lookup, which "fails" then falls back to the upstream gsw implementation. As far as I can tell, "density" needs to both exist as ``density.py`` for the import machinery, and as a name the ``__init__.py`` is aware of (so it does the right thing).

I could not figure out a way that reliably tests this with all the import caching going on.

**Comment by rcaneill** [comment]
I never reviewed a PR, I guess I should clone your branch locally (I need to find how to do this), compute the tests, have a look at the code, and verify that everything works fine?

**Comment by rcaneill** [comment]
I added some extra tests to test all possibilities of import, unfortunately using:
```python
from gsw.density import sigma0
sigma0(...)
```
or
```python
from gsw import density
density.sigma0
```

both fall on the gsw backend implementation.
Maybe it has something to do with the caching of imports, or something more complicated.

Loading every function from `gsw.density` does not import anything actually...
```python
from gsw.density import * # Not good to do, but checks if functions are loaded
sigma1
```

For the import caching you mentioned, if we implement the submodule tests in different test files, should reset the cache no? Or were you speaking about the '.pyc' caches?

I don't see any solution to be able to import in whatever way exists yet...

**Comment by DocOtak** [comment]
Good review! :grin:

Some things going on:

At the top of the tests, we have a:
```python
import gsw_xarray as gsw
```

This allows you to directly access the functions like:

```python
gsw.sigma0 # is is calling the renamed gsw_xarray from above
```

When you do the following:

```python
gsw.density.sigma0
```

Since that `gsw` is a renamed gsw_xarray, the `.density` becomes an attribute lookup, the `__getattr__` is called to try to find the name. Initially we only had the actual functions as valid name, so the lookup failed, and then upstream gsw library is searched for the name (which succeeds).

However, when you do:
```python
from gsw_xarray.density import sigma0
```
The module loader machinery is doing that lookup, so the `__getattr__` method is ignored and whatever loader (this means something specific in the importlib) that was returned is doing its thing (I don't know details).

This also means the following:
```python
import gsw_xarray as gsw
gsw.density.sigma0 # will be our wrapped version

from gsw.density import sigma0 # uses the import machinery, will always find the upstream gsw module
```

Fixing the `from gsw_xarray.density import *` can be done by adding the correct `__all__` value to each file.

This PR needs some more work...

**Comment by rcaneill** [comment]
I re-wrote the module tests in a new file. I also split them into different functions to test 1 (or 2) feature at a time.
I had some time to understand all the machinery you did Andrew, and now I understand how these `__dir__` and `__getattr__` functions work more or less.

If I summarize, I think that gsw_xarray should allow the following imports, with a preference for the 1st two methods:
1. `import gsw_xarray as gsw` and then use directly the functions, e.g. `gsw.sigma0`
2. `from gsw_xarray import sigma0`
3. `from gsw_xarray.density import sigma0`
4. `from gsw_xarray import density`, and use `density.sigma0`
5. [can be discussed] `from gsw_xarray import *`
6. [can be discussed] `from gsw_xarray.density import *`

The 4 first methods are now working and have their own tests in `test_imports.py`. The 5th adn 6th methods would need to use `__all__ = __dir__()` I guess(?), or something similar.

Now I also understand that the tests I wrote previously were falling on the gsw backend because I were importing the functions from gsw directly, as you answered me...
> I added some extra tests to test all possibilities of import, unfortunately using:
> 
> ```python
> from gsw.density import sigma0
> sigma0(...)
> ```
> 
> or
> 
> ```python
> from gsw import density
> density.sigma0
> ```
> 
> both fall on the gsw backend implementation. Maybe it has something to do with the caching of imports, or something more complicated.


Tell me if I missed something.

**Review by rcaneill — APPROVED** [review]
Seems good to me!
You should also review my work before we can merge.

**Comment by DocOtak** [comment]
Was away for a few days, will check in on this. I have some ideas about testing the API surface against the upstream gsw that I'd like to try out before merging. That said, this feels like it is holding up other tasks.

Thanks for getting milestones started!

**Comment by DocOtak** [comment]
One more look would be awesome.

Some explanation of changes to the tests:
* I turned the test dataset function into a pytest fixture. So now you just pass in the name "ds" into a test function that needs it and pytest does magic.
* Extract all the names from upstream gsw and ensure we emulate all of them
* Ensure that we return wrapped versions of functions we have wrappers for
* Do the 2 above for all the submodules in upstream gsw
* 2 import style checks done using exec into a test specific "locals" dict
  * The importlib functions can only find modules, but we want to check individual function names being imported, hence the use of "exec"
* a test to ensure access in the form e.g. `gsw.density.sigma0` work as expected
* The above tests also check for the negative constraint, that submodules raise an Attribute or Import error when a name does not exist in the upstream.

On the todo, is to make the wildcard imports work: e.g. `from gsw_xarray.density import *`, but at this point I'd want to handle it in a new PR.

File: gsw_xarray/tests/conftest.py
Comment (rcaneill) [inline gsw_xarray/tests/conftest.py:7]: This functionality is very nice :)

**Review by rcaneill — APPROVED** [review]
Nice changes on the tests! I went through them, it seems correct to me. Very smart way to solve the problem of all cases, for all functions!

File: gsw_xarray/tests/test_imports.py
Comment (DocOtak) [reply gsw_xarray/tests/test_imports.py:114]: The upstream gsw doesn't directly expose the actual c compiled ufuncs, it wrapps them all in python funcs of the same name (with some decorators and doc strings). There are a few places in the upstream library that directly use the c ufuncs internally. Due to the way it's imported in those places, the name is exposed and importable.

So you are correct in that `gsw.stability.grav` is not the same as `gsw.grav` in the upstream. This could potentially be solved by only looking at what is in `__all__` but that really only matters when doing a wildcard import.

Comment (DocOtak) [reply gsw_xarray/tests/test_imports.py:114]: Basically... the upstream leaks its private api a little. Not knowing what to do for these cases, I just xfailed them. Maybe we can raise with the other gsw-python maintainers.

**Pull request merged**
